core/src/main/java/org/elasticsearch/script/Script.java

```diff
@@ -32,268 +31,545 @@
-     *                      script source code. For indexed scripts this is the id used in the request. For on file
-     *                      scripts this is the file name.
-     * @param type          The type of script -- dynamic, stored, or file.
-     * @param lang          The language of the script to be compiled/executed.
-     * @param params        The map of parameters the script will be executed with.
-     * @param contentType   The {@link XContentType} of the script. Only relevant for inline scripts that have not been
-     *                      defined as a plain string, but as json or yaml content. This class needs this information
-     *                      when serializing the script back to xcontent.
+     * "return Math.log(doc.popularity) * 100;"
+     * }
+     *
+     * The complex format where {@link ScriptType} and idOrCode are required while lang, options and params are not required.
+     *
+     * {@code
+     * {
+     *     "<type (inline, stored, file)>" : "<idOrCode>",
+     *     "lang" : "<lang>",
+     *     "options" : {
+     *         "option0" : "<option0>",
+     *         "option1" : "<option1>",
+     *         ...
+     *     }
+     *     "params" : {
+     *         "param0" : "<param0>",
+     *         "param1" : "<param1>",
+     *         ...
+     *     }
+     * }
+     *
+     * {
+     *     "inline" : "return Math.log(doc.popularity) * params.multiplier",
+     *     "lang" : "painless",
+     *     "params" : {
+     *         "multiplier" : 100.0
+     *     }
+     * }
+     * }
+     *
+     * This also handles templates in a special way.  If a complexly formatted query is specified as another complex
+     * JSON object the query is assumed to be a template, and the format will be preserved.
+     *
+     * {@code
+     * {
+     *     "inline" : { "query" : ... },
+     *     "lang" : "<lang>",
+     *     "options" : {
+     *         "option0" : "<option0>",
+     *         "option1" : "<option1>",
+     *         ...
+     *     }
+     *     "params" : {
+     *         "param0" : "<param0>",
+     *         "param1" : "<param1>",
+     *         ...
+     *     }
+     * }
+     * }
+     *
+     * @param parser  The {@link XContentParser} to be used.
+     * @param matcher The {@link ParseFieldMatcher} to be used.
+     * @param lang    The default language to use if no language is specified.  The default language isn't necessarily
+     *                the one defined by {@link Script#DEFAULT_SCRIPT_LANG} due to backwards compatiblity requirements
+     *                related to stored queries using previously default languauges.
+     * @return        The parsed {@link Script}.
      */
+    public static Script parse(XContentParser parser, ParseFieldMatcher matcher, String lang) throws IOException {
+        Objects.requireNonNull(lang);
+
+        Token token = parser.currentToken();
```

Comment:
@s1monw I would like to do this, but there's a couple of things that are not obvious to me if they can be done or not.  So I have the following questions:

1) Templates can be parsed as either a String or an Object (that needs to be preserved), but use the same parse field for this -- ```"inline" : { "query" : }...``` or ```"inline" : "{ \"query\": ...}``` -- is there a way to specify in a ConstructingObjectParser that a parse field can be multiple types?
2) If a lang is not provided I need to use a default, but that default isn't necessarily a constant for back-compat :/ , so is there a way to pass in an additional parameter that wasn't parsed?
3) This can be solved by doing a couple of hack casts, but the compiler options is currently a Map<String, String>, and the parser only supports Map<String, Object> as far as I can tell.  This would mean if we do open up some compiler options in the future, this could lead to a potentially bad error.  Is there a way to enforce the Map<String, String> appropriately without having to retroactively go back through and check each item in the Map after it's been parsed?